### PR TITLE
Adds initial file upload and management.

### LINKED ISSUE #141: Basic file upload

### PR DESCRIPTION
**Pull request by justinlittman**
closes #141

https://github.com/user-attachments/assets/06812c93-83a6-4f85-b1c6-627eabdd7842

### CODE REVIEW

**Comment by justinlittman** [comment]
Follow-up work:
* Display files on work show page.
* Pagination
* Clean up job for Content / Content Files
* Authorization for Content / Content files controllers
* Upload UI enhancements

File: app/controllers/content_files_controller.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  def show; end
+
+  def edit; end
+
+  def update
+    if @content_file.update(content_file_params)
+      redirect_to content_file_path(@content_file)
+    else
+      render :edit, status: :unprocessable_entity
```
Comment (mjgiarlo) [inline app/controllers/content_files_controller.rb:17]: Shows as uncovered. Worth adding a spec to test the unhappy path?

File: app/controllers/content_files_controller.rb [hunk 2]
```diff
@@ -0,0 +1,35 @@
+# frozen_string_literal: true
+
+# Controller for a Work content file
+class ContentFilesController < ApplicationController
+  # Skipping authorization for now
```
Comment (mjgiarlo) [inline app/controllers/content_files_controller.rb:5]: Could you prepend a `TODO: ` here and/or create a ticket if we don't have one?

File: app/controllers/contents_controller.rb [hunk 3]
```diff
@@ -0,0 +1,50 @@
+# frozen_string_literal: true
+
+# Controller for a Work contents (files)
+class ContentsController < ApplicationController
+  # Skipping authorization for now
```
Comment (mjgiarlo) [inline app/controllers/contents_controller.rb:5]: Same comment as above re: todo/ticketing

File: app/controllers/contents_controller.rb [hunk 4]
```diff
@@ -0,0 +1,50 @@
...
+
+  def update
+    if params.dig(:content, :files).present?
+      update_files
+      return respond_to do |format|
+        format.turbo_stream
+      end
+    end
+
+    redirect_to edit_content_path(@content)
```
Comment (mjgiarlo) [inline app/controllers/contents_controller.rb:23]: Also uncovered. Worth testing?

File: app/controllers/contents_controller.rb [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+    @content = Content.find(params[:id])
+  end
+
+  # ActionDispath::Http::UploadedFile only provides the base name.
+  # This attempts to get the complete filename from the headers.
+  def filename_for(file)
+    if (matcher = file.headers.match(/filename="(.+)"/))
+      matcher[1]
+    else
+      file.original_filename
```
Comment (mjgiarlo) [inline app/controllers/contents_controller.rb:38]: Also uncovered. Worth testing?

File: app/controllers/works_controller.rb [hunk 6]
```diff
@@ -62,6 +63,7 @@ def update
       DepositJob.perform_later(work: @work, work_form: @work_form, deposit: deposit?)
       redirect_to wait_works_path(@work.id)
     else
+      @content = Content.find(@work_form.content_id)
```
Comment (mjgiarlo) [inline app/controllers/works_controller.rb:66]: Also uncovered. Worth testing?

File: app/services/contents/builder.rb [hunk 7]
```diff
@@ -7,13 +7,14 @@ def self.call(...)
       new(...).call
```
Comment (mjgiarlo) [inline app/services/contents/builder.rb:7]: I'm fuzzy on how we're distinguishing mappers from builders. What do you see as the difference between them?

File: app/views/content_files/_show.html.erb [hunk 8]
```diff
@@ -0,0 +1,22 @@
+<tr>
```
Comment (mjgiarlo) [inline app/views/content_files/_show.html.erb:1]: Trivial, but: should we stick to two-space indents in ERBs? I fear our editors are going to keep introducing large whitespace diffs, but maybe that's no big deal.

File: app/views/works/form.html.erb [hunk 9]
```diff
@@ -13,8 +14,18 @@
...
       <%= form.hidden_field :content_id %>
       <div class="tab-content h-100">
-        <%= render Works::Edit::PaneComponent.new(tab_name: :title, label: I18n.t('works.edit.panes.title.label'), form:, selected: true) do %>
-          <%= render Elements::Forms::TextFieldComponent.new(form:, field_name: :title, required: true, label: I18n.t('works.edit.fields.title.label'),
+        <%= render Works::Edit::PaneComponent.new(tab_name: :files, form:, label: I18n.t('works.edit.panes.files.label'), selected: true) do %>
+          <%= turbo_frame_tag @content, 'edit', src: edit_content_path(@content) do %>
+              <%= render Elements::SpinnerComponent.new %>
+          <% end %>
+          <%= render Elements::HorizontalRuleComponent.new(classes: 'my-3') %>
+          <div class="h5 mb-3">Your files</div>
```
Comment (mjgiarlo) [inline app/views/works/form.html.erb:22]: Worth i18n-ing?

File: app/controllers/content_files_controller.rb (hunk 1 above)
Comment (justinlittman) [reply app/controllers/content_files_controller.rb:17]: At this point, I don't know how I would I would test, as Content File doesn't have any validation could fail at this point. I'm putting it in proactively should we add validation in the future.

Thoughts?

File: app/controllers/content_files_controller.rb (hunk 2 above)
Comment (justinlittman) [reply app/controllers/content_files_controller.rb:5]: https://github.com/sul-dlss/hungry-hungry-hippo/issues/152

File: app/controllers/contents_controller.rb (hunk 4 above)
Comment (justinlittman) [reply app/controllers/contents_controller.rb:23]: I actually don't think this is necessary, so removing.

File: app/controllers/contents_controller.rb (hunk 5 above)
Comment (justinlittman) [reply app/controllers/contents_controller.rb:38]: This seems unnecessary. Removing.

File: app/controllers/content_files_controller.rb (hunk 1 above)
Comment (mjgiarlo) [reply app/controllers/content_files_controller.rb:17]: I figure you could stub `update` to return `false` and validate that the form is reloaded? If that's more futzing than value, we can skip it.

File: app/controllers/works_controller.rb (hunk 6 above)
Comment (justinlittman) [reply app/controllers/works_controller.rb:66]: Adding testing.

File: app/services/contents/builder.rb (hunk 7 above)
Comment (justinlittman) [reply app/services/contents/builder.rb:7]: I'm my mind, a builder creates AR objects, whereas a mapper just maps without persisting anything.

But I don't feel strongly about this if you have other thoughts.

File: app/views/content_files/_show.html.erb (hunk 8 above)
Comment (justinlittman) [reply app/views/content_files/_show.html.erb:1]: Fixed.

File: app/views/works/form.html.erb (hunk 9 above)
Comment (justinlittman) [reply app/views/works/form.html.erb:22]: Yup.

File: app/services/contents/builder.rb (hunk 7 above)
Comment (mjgiarlo) [reply app/services/contents/builder.rb:7]: That's a good distinction. I don't care too much whether we call our builder classes "builders" or "data mappers" or whatever, as long as we developers can keep the distinction in our heads. I might do a short spike on teasing these apart, and we can either keep it or throw it away.

**Review by mjgiarlo — APPROVED** [review]
Looking good. Thanks!

**Pull request merged**